File: day10/day10_2.rb
adapters = ARGF.readlines.map { |line| line.chomp.to_i }.sort
adapters.unshift(0)
adapters.push(adapters.last + 3)
adapters = adapters.each.with_index.to_a

def traverse(adapters, current_idx, table)
  # BASE CASE
  if current_idx == adapters.count - 1
    return table[current_idx] = 1
  end

  current = adapters[current_idx][0]
  to_search = adapters[current_idx+1..].take_while { |(adapter, idx)| adapter - 3 <= current }

  table[current_idx] = to_search.sum do |(adapter, idx)|
    table[idx] ? table[idx] : traverse(adapters, idx, table)
  end
end

p "Solution: #{traverse(adapters, 0, {})}"

File: 2020/10/python/day10_2.py
import sys


def traverse(current_idx=0, table={}):
    if current_idx == len(adapters) - 1:
        table[current_idx] = 1
        return 1

    current = adapters[current_idx]
    option_idxs = range(current_idx + 1, current_idx + 4)
    to_search = [i for i in option_idxs if i < len(
        adapters) and adapters[i] - 3 <= current]

    sum = 0
    for search_idx in to_search:
        if search_idx in table:
            sum += table[search_idx]
            continue
        sum += traverse(search_idx, table)

    table[current_idx] = sum
    return table[current_idx]


if __name__ == "__main__":
    adapters = sorted(
        list(map(lambda s: int(s.rstrip()), sys.stdin.readlines())))
    adapters.insert(0, 0)
    adapters.append(adapters[-1] + 3)

    print(traverse())
